### PR TITLE
feat: add benchmarks for BTreeMap

### PR DESCRIPTION
**Pull request by ielashi**
# Problem
The benchmarks we have aren't comprehensive enough to get an accurate picture of the performance of `BTreeMap`. For example, in the `0.4.0` release, the benchmarking numbers showed us an improvement of ~40% in the `insert` and `remove` operations, but when I profiled the Bitcoin canister with that upgrade I saw a _regression_ of about 40%.

# Solution
Add more thorough benchmarks for `insert`, `remove`, and `get` to get a more accurate picture of the performance impact of our changes. These will then be used to address the performance regressions we've seen with `0.4.0+`

### CODE REVIEW

File: benchmarks/run.sh [hunk 1]
```diff
@@ -12,14 +12,40 @@ trap "dfx stop" ERR EXIT
...
+
+dfx canister call benchmarks btreemap_get_blob_4_1024 --query
+dfx canister call benchmarks btreemap_get_blob_8_1024 --query
+dfx canister call benchmarks btreemap_get_blob_16_1024 --query
+dfx canister call benchmarks btreemap_get_blob_32_1024 --query
+dfx canister call benchmarks btreemap_get_blob_64_1024 --query
+dfx canister call benchmarks btreemap_get_blob_128_1024 --query
+# These tests go over the instruction limit, so we can't run them currently.
+#dfx canister call benchmarks btreemap_get_blob_256_1024
+#dfx canister call benchmarks btreemap_get_blob_512_1024
```
Comment (ielashi) [inline benchmarks/run.sh:40]: In a future PR, I plan to use a version of `drun` to allow us to run these benchmarks without a cycles limit.

Comment (dsarlis) [reply benchmarks/run.sh:40]: They go over even with the increased DTS limits?

File: benchmarks/src/btreemap.rs [hunk 2]
```diff
@@ -1,47 +1,252 @@
...
+use ic_cdk_macros::{query, update};
+use ic_stable_structures::{storable::Blob, BTreeMap, BoundedStorable, DefaultMemoryImpl};
 use tiny_rng::{Rand, Rng};
 
-/// Benchmarks inserting keys into a BTreeMap.
-#[ic_cdk_macros::query]
-pub fn btreemap_insert() -> u64 {
-    let mut btree: BTreeMap<u64, (), _> = BTreeMap::init(DefaultMemoryImpl::default());
+#[query]
+pub fn btreemap_insert_blob_4_1024() -> u64 {
```
Comment (dsarlis) [inline benchmarks/src/btreemap.rs:7]: nit: We can probably reduce all this boilerplate with a simple macro that defines the functions for each combination.

**Review by dsarlis — APPROVED** [review]
> For example, in the 0.4.0 release, the benchmarking numbers showed us an improvement of ~40% in the insert and remove operations, but when I profiled the Bitcoin canister with that upgrade I saw a regression of about 40%.

Ouch...

File: benchmarks/run.sh (hunk 1 above)
Comment (ielashi) [reply benchmarks/run.sh:40]: The replica that ships with 0.12.1 has a limit of 10B, and that's enough for us at the moment.

File: benchmarks/src/btreemap.rs (hunk 2 above)
Comment (ielashi) [reply benchmarks/src/btreemap.rs:7]: True. This code will evolve quite a bit in the next commits though, so I want to first wait until the code is more stable before working on removing the boilerplate.

**Pull request merged**